solver/src/settlement.rs

```diff
@@ -280,20 +295,40 @@ fn sell_order_surplus(
     buy_token_price: &BigRational,
     sell_amount_limit: &BigRational,
     buy_amount_limit: &BigRational,
-    executed_amount: &BigRational,
+    executed_sell_amount: &BigRational,
 ) -> Option<BigRational> {
     if sell_amount_limit.is_zero() {
         return None;
     }
-    let res = executed_amount * sell_token_price
-        - (executed_amount * buy_amount_limit / sell_amount_limit * buy_token_price);
+    // absolute_surplus = exec_buy_amount - (exec_sell_amount * limit_buy_amount/limit_sell_amount)
+    // external_price(buy_token) * absolute_surplus
+    // -------------
+    let limit_buy_amount = executed_sell_amount * buy_amount_limit / sell_amount_limit;
+    let res = (executed_sell_amount * sell_token_price) - (limit_buy_amount * buy_token_price);
     if res.is_negative() {
         None
     } else {
         Some(res)
     }
 }
 
+fn surplus_ratio(
+    sell_token_price: &BigRational,
+    buy_token_price: &BigRational,
+    sell_amount_limit: &BigRational,
+    buy_amount_limit: &BigRational,
+) -> Option<BigRational> {
+    if buy_token_price.is_zero() {
```

Comment:
I think you need to check that both `buy_token_price` and `buy_amount_limit` are not `0` or you could panic.